node/src/test_utils/recorder_stop_conditions.rs

```diff
@@ -0,0 +1,325 @@
+            }
+            StopConditions::All(_) => None,
+        }
+    }
+
+    fn resolve_any<T: PartialEq + Send + 'static>(
+        conditions: &Vec<StopCondition>,
+        msg: &T,
+    ) -> bool {
+        conditions
+            .iter()
+            .any(|condition| condition.resolve_condition::<T>(msg))
+    }
+
+    fn inspect_mutable<T: PartialEq + Send + 'static>(&mut self, msg: &T) -> bool {
+        match self {
+            StopConditions::All(conditions) => {
+                let indexes_to_remove = Self::indexes_of_matched_conditions(conditions, msg);
+                Self::remove_matched_conditions(conditions, indexes_to_remove);
+                conditions.is_empty()
+            }
+            _ => unreachable!("something is wrong"),
+        }
+    }
+
+    fn indexes_of_matched_conditions<T: PartialEq + Send + 'static>(
+        conditions: &[StopCondition],
+        msg: &T,
+    ) -> Vec<usize> {
+        conditions
+            .iter()
+            .enumerate()
+            .fold(vec![], |mut acc, (idx, condition)| {
+                let matches = condition.resolve_condition::<T>(msg);
+                if matches {
+                    acc.push(idx)
+                }
+                acc
+            })
+    }
+
+    fn remove_matched_conditions(
+        conditions: &mut Vec<StopCondition>,
+        indexes_to_remove: Vec<usize>,
+    ) {
+        if !indexes_to_remove.is_empty() {
+            let _ = indexes_to_remove
+                .into_iter()
+                .fold(0, |removed_counter: usize, idx| {
+                    conditions.remove(idx - removed_counter);
+                    removed_counter + 1
+                });
+        }
+    }
+}
+
+impl StopCondition {
+    fn resolve_condition<T: PartialEq + Send + 'static>(&self, msg: &T) -> bool {
+        match self {
+            StopCondition::StopOnType(type_id) => Self::matches_stop_on_type::<T>(*type_id),
+            StopCondition::StopOnMatch { exemplar } => {
+                Self::matches_stop_on_match::<T>(exemplar, msg)
+            }
+            StopCondition::StopOnPredicate { predicate } => {
+                Self::matches_stop_on_predicate(predicate.as_ref(), msg)
+            }
+        }
+    }
+
+    fn matches_stop_on_type<T: 'static>(expected_type_id: TypeId) -> bool {
+        let msg_type_id = TypeId::of::<T>();
+        msg_type_id == expected_type_id
+    }
+
+    fn matches_stop_on_match<T: PartialEq + 'static + Send>(
+        exemplar: &BoxedMsgExpected,
+        msg: &T,
+    ) -> bool {
+        if let Some(downcast_exemplar) = exemplar.downcast_ref::<T>() {
+            return downcast_exemplar == msg;
+        }
+        false
+    }
+
+    fn matches_stop_on_predicate<T: Send + 'static>(
+        predicate: &dyn Fn(RefMsgExpected) -> bool,
+        msg: &T,
+    ) -> bool {
+        predicate(msg as RefMsgExpected)
+    }
+}
+
+#[macro_export]
+macro_rules! single_type_id {
```

Comment:
For consistency, this macro name should change as well, perhaps to `match_type_id!`.